### PR TITLE
Revoke old access_token upon force refresh

### PR DESCRIPTION
**Pull request by vahancouver**
[SDKS-2571](https://bugster.forgerock.org/jira/browse/SDKS-2571)

Revoke old access_token upon force refresh. Made sure it's only revoking the access token and not the refresh token.

**The following regression test has been completed (this test has been verified with "Issue Refresh Tokens on Refreshing Access Tokens" both On and Off in AM console)**

1. Call Login - Verify that access token and refresh token has been issued

2. Call Get User Info - Verify that user info successfully retrieved using the Access Token from Step 1

3. Call Refresh Token - Verify that the new Access Token has been issued (check that it's different from the Access Token from Step 1) and that Revoke endpoint has been called with the old Access Token from Step 1

4. Call Get User Info - Verify that user info successfully retrieved using the Access Token from Step 3

5. In the Terminal call /userinfo endpoint using the Access Token from Step 1 - Verify that "invalid_token" response has been received

### CODE REVIEW

**Comment by jeyanthanperiyasamy** [comment]
Can you do a regression test like this by mocking all the endpoints ? 

1. GetAccessToken
2. Call RefreshToken
3. verify the revoke endpoint called with the previous token
4.  verify the previous token is not available in the cache

**Review by jeyanthanperiyasamy — COMMENTED** [review]
comments added

File: FRAuth/FRAuth/Manager/TokenManager.swift [hunk 1]
```diff
@@ -365,4 +365,12 @@ struct TokenManager {
         FRUser._staticUser = nil
         Browser.currentBrowser = nil
     }
+    
+    
+    /// Revoke given token
+    func revokeToken(_ token: AccessToken, completion: @escaping CompletionCallback) {
+        self.oAuth2Client.revoke(accessToken: token) { (error) in
```
Comment (george-bafaloukas-forgerock) [inline FRAuth/FRAuth/Manager/TokenManager.swift:372]: @vahancouver I think this is not going to work.
Calling this will call:
`public func revoke(accessToken: AccessToken, completion: @escaping CompletionCallback) {
        // Construct parameter for the request
        var parameter:[String: String] = [:]
        let token = accessToken.refreshToken ?? accessToken.value
        parameter[OAuth2.token] = token
        parameter[OAuth2.clientId] = self.clientId
        
        let request = Request(url: self.serverConfig.tokenRevokeURL, method: .POST, headers: [:], bodyParams: parameter, urlParams: [:], requestType: .urlEncoded, responseType: .json, timeoutInterval: self.serverConfig.timeout)
        
        FRRestClient.invoke(request: request, action: Action(type: .REVOKE_TOKEN)) { (result) in
            switch result {
            case .success(_ , _):
                completion(nil)
                break
            case .failure(let error):
                completion(error)
                break
            }
        }
    }
    `

Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Manager/TokenManager.swift:372]: That code will revoke based on the Refresh token, by default and will thus revoke all accessTokens created by this refresh token. And the refresh token as well. Unless mistaken, this will end up saving a token that has been revoked

**Review by george-bafaloukas-forgerock — CHANGES_REQUESTED** [review]

**Comment by george-bafaloukas-forgerock** [comment]
As a general note, I would think we would need to change the `OAuth2Client` revoke method to take an input whether to revoke using the refresh token or the accessToken

**Review by george-bafaloukas-forgerock — APPROVED** [review]
Looks good. Make sure you test the edge cases with the Refresh tokens please

**Review by spetrov — APPROVED** [review]
LGTM! 👍🏻
